### ISSUE TITLE
KeyError: 'Failed to format this callback filepath: "ckpt_dir/{epoch:02d}-{val_loss:.2f}". Reason: \'val_loss\''

### ISSUE THREAD

**Issue by zeeros** [BNN-UPC/RouteNet-Erlang#3]
Training is failing at the second epoch in `scalability/main.py`

**Epoch 1**
```
Epoch 1/200
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_9_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_9_grad/Reshape:0", dtype=float32), dense_shape=Tensor("gradients/GatherV2_9_grad/Cast:0", shape=(None,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_8_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_8_grad/Reshape:0", shape=(None, 1), dtype=float32), dense_shape=Tensor("gradients/GatherV2_8_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_7_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_7_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_7_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_6_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_6_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_6_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_5_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_5_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_5_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_4_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_4_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_4_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_3_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_3_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_3_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_2_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_2_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_2_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_1_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_1_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_1_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
/anaconda/envs/azureml_py38/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradients/GatherV2_grad/Reshape_1:0", shape=(None,), dtype=int32), values=Tensor("gradients/GatherV2_grad/Reshape:0", shape=(None, 32), dtype=float32), dense_shape=Tensor("gradients/GatherV2_grad/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
  warnings.warn(
2022-04-16 20:41:49.784273: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
2022-04-16 20:42:05.646761: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 26 of 1000
2022-04-16 20:42:15.469365: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 55 of 1000
2022-04-16 20:42:25.503645: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 94 of 1000
2022-04-16 20:42:35.598784: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 134 of 1000
2022-04-16 20:42:45.273349: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 174 of 1000
2022-04-16 20:42:55.381876: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 223 of 1000
2022-04-16 20:43:05.389659: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 280 of 1000
^[>2022-04-16 20:43:15.323200: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 314 of 1000
2022-04-16 20:43:25.298741: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 350 of 1000
2022-04-16 20:43:35.351326: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 426 of 1000
2022-04-16 20:43:45.920267: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 466 of 1000
2022-04-16 20:43:55.303526: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 513 of 1000
2022-04-16 20:44:05.397849: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 546 of 1000
2022-04-16 20:44:15.733292: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 597 of 1000
2022-04-16 20:44:25.349787: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 652 of 1000
2022-04-16 20:44:35.425504: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 723 of 1000
2022-04-16 20:44:45.438521: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 769 of 1000
2022-04-16 20:44:55.287286: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 812 of 1000
2022-04-16 20:45:05.381177: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 873 of 1000
2022-04-16 20:45:15.400310: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 935 of 1000
2022-04-16 20:45:25.343842: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 974 of 1000
2022-04-16 20:45:28.238022: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
4000/4000 [==============================] - ETA: 0s - loss: 24.3972WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 20 batches). You may need to use the repeat() function when building your dataset.
4000/4000 [==============================] - 1059s 208ms/step - loss: 24.3972 - val_loss: 24.5493

Epoch 00001: saving model to ckpt_dir/01-24.55
```

**Epoch 2**
```
Epoch 2/200
4000/4000 [==============================] - 632s 158ms/step - loss: 10.6697
Traceback (most recent call last):
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/callbacks.py", line 1458, in _get_file_path
    file_path = self.filepath.format(epoch=epoch + 1, **logs)
KeyError: 'val_loss'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 92, in <module>
    model.fit(ds_train,
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/engine/training.py", line 1230, in fit
    callbacks.on_epoch_end(epoch, epoch_logs)
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/callbacks.py", line 413, in on_epoch_end
    callback.on_epoch_end(epoch, logs)
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/callbacks.py", line 1368, in on_epoch_end
    self._save_model(epoch=epoch, batch=None, logs=logs)
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/callbacks.py", line 1403, in _save_model
    filepath = self._get_file_path(epoch, batch, logs)
  File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/keras/callbacks.py", line 1463, in _get_file_path
    raise KeyError('Failed to format this callback filepath: "{}". '
KeyError: 'Failed to format this callback filepath: "ckpt_dir/{epoch:02d}-{val_loss:.2f}". Reason: \'val_loss\''
```

Packages installed

```
Name: tensorflow
Version: 2.6.3
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /anaconda/envs/azureml_py38/lib/python3.8/site-packages
Requires: typing-extensions, tensorboard, wheel, google-pasta, h5py, absl-py, flatbuffers, keras-preprocessing, opt-einsum, gast, protobuf, astunparse, numpy, grpcio, termcolor, clang, keras, wrapt, six, tensorflow-estimator
Required-by: autokeras
---
Name: networkx
Version: 2.5
Summary: Python package for creating and manipulating graphs and networks
Home-page: http://networkx.github.io/
Author: Aric Hagberg
Author-email: hagberg@lanl.gov
License: UNKNOWN
Location: /anaconda/envs/azureml_py38/lib/python3.8/site-packages
Requires: decorator
Required-by: visions, scikit-image, responsibleai, dowhy
---
Name: configparser
Version: 5.2.0
Summary: Updated configparser from Python 3.8 for Python 2.6+.
Home-page: https://github.com/jaraco/configparser/
Author: Łukasz Langa
Author-email: lukasz@langa.pl
License: UNKNOWN
Location: /anaconda/envs/azureml_py38/lib/python3.8/site-packages
Requires: 
Required-by: azureml-defaults
---
Name: pandas
Version: 1.3.5
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: The Pandas Development Team
Author-email: pandas-dev@python.org
License: BSD-3-Clause
Location: /anaconda/envs/azureml_py38/lib/python3.8/site-packages
Requires: python-dateutil, pytz, numpy
Required-by: visions, torch-tb-profiler, statsmodels, sklearn-pandas, shap, seaborn, scrapbook, responsibleai, raiwidgets, pyLDAvis, pycaret, pmdarima, phik, pandas-profiling, pandas-ml, nimbusml, mlxtend, mlflow, ml-wrappers, interpret-community, fbprophet, fastparquet, fastai, fairlearn, erroranalysis, econml, dowhy, dice-ml, datasets, dask-sql, cufflinks, azureml-training-tabular, azureml-train-automl-runtime, azureml-opendatasets, azureml-datadrift, azureml-automl-runtime, autokeras, arch
```

**Comment by zeeros** [comment]
An easy fix is to just remove the loss value from the path

`filepath = os.path.join(ckpt_dir, "{epoch:02d}")#-{val_loss:.2f}")`

**Comment by zeeros** [comment]
The problem might be related to the validation data, in fact I see that at the second epoch the `val_loss` is not available.

https://stackoverflow.com/a/56849050/2265222

**Comment by MiquelFerriol** [comment]
Dear @zeeros,
Thank you for your interest in our work and for pointing out this problem.

As you mention, this error is related to the validation dataset that is declared at [line46](https://github.com/BNN-UPC/RouteNet-Erlang/blob/73fd3da812b80f4e2079ea60d308413c3a52b396/Scalability/main.py#L46)
`ds_test = input_fn('../data/scalability/test/300', min_scale=10, max_scale=11, shuffle=False)`

The problem is that there are not enough samples at the `../data/scalability/test/300` directory, so, at the second epoch, the generator cannot "yield" more. This causes the callback to fail since there is no "val_loss" returned by the validation process.

The error should be fixed at #6 by simply changing the directory of the ds_test declaration by:
`ds_test = input_fn('../data/scalability/test/', min_scale=10, max_scale=11, shuffle=False)`

Note that, if you want to validate over a specific topology size during the training phase, you can do so by specifying the topology size as it was done right now. However, you will probably need to reduce the `validation_steps` of the [fit function](https://github.com/BNN-UPC/RouteNet-Erlang/blob/73fd3da812b80f4e2079ea60d308413c3a52b396/Scalability/main.py#L92) to avoid this problem.

**Comment by zeeros** [comment]
I confirm that #6 fixes the issue, and thanks for explaining its context.
